File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/java/src/main/java/com/example/greedy/algorithms/FractionalKnapsack.java
package com.example.greedy.algorithms;

import com.example.greedy.model.Item;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public final class FractionalKnapsack {
    private FractionalKnapsack() {
    }

    public static final class Take {
        private final String name;
        private final double fraction;
        private final double value;

        public Take(String name, double fraction, double value) {
            this.name = name;
            this.fraction = fraction;
            this.value = value;
        }

        @Override
        public String toString() {
            return "Take{name='" + name + "', fraction=" + fraction + ", value=" + value + "}";
        }
    }

    public static final class Result {
        private final double totalValue;
        private final List<Take> taken;

        public Result(double totalValue, List<Take> taken) {
            this.totalValue = totalValue;
            this.taken = List.copyOf(taken);
        }

        public double totalValue() {
            return totalValue;
        }

        @Override
        public String toString() {
            return "Result{totalValue=" + totalValue + ", taken=" + taken + "}";
        }
    }

    public static Result solve(List<Item> items, double capacity) {
        if (capacity < 0) {
            throw new IllegalArgumentException("capacity must be non-negative");
        }

        double remaining = capacity;
        double totalValue = 0;
        List<Take> taken = new ArrayList<>();

        List<Item> sorted = items.stream()
                .sorted(Comparator.comparingDouble(Item::density).reversed())
                .collect(Collectors.toList());

        for (Item item : sorted) {
            if (remaining <= 0) {
                break;
            }
            double amount = Math.min(item.weight(), remaining);
            double fraction = amount / item.weight();
            double value = item.value() * fraction;
            taken.add(new Take(item.name(), round(fraction), round(value)));
            totalValue += value;
            remaining -= amount;
        }

        return new Result(round(totalValue), taken);
    }

    private static double round(double value) {
        return Math.round(value * 100.0) / 100.0;
    }
}

File: 29Items-Projects/7Algorithms/4-Greedy-algorithms/python/greedy_algorithms/algorithms/fractional_knapsack.py
from dataclasses import dataclass


@dataclass(frozen=True)
class Item:
    name: str
    value: float
    weight: float

    @property
    def density(self) -> float:
        if self.weight <= 0:
            raise ValueError(f"{self.name} must have positive weight")
        return self.value / self.weight


@dataclass(frozen=True)
class KnapsackTake:
    name: str
    fraction: float
    value: float


def fractional_knapsack(items: list[Item], capacity: float) -> dict[str, object]:
    if capacity < 0:
        raise ValueError("capacity must be non-negative")

    remaining = capacity
    total_value = 0.0
    taken: list[KnapsackTake] = []

    for item in sorted(items, key=lambda candidate: candidate.density, reverse=True):
        if remaining <= 0:
            break
        amount = min(item.weight, remaining)
        fraction = amount / item.weight
        value = item.value * fraction
        taken.append(KnapsackTake(item.name, round(fraction, 2), round(value, 2)))
        total_value += value
        remaining -= amount

    return {"total_value": round(total_value, 2), "taken": taken}
